File: smtp/typescript/examples/attachments.ts
/**
 * Email with Attachments Example
 *
 * This example demonstrates how to send emails with file attachments
 * and inline images embedded in HTML content.
 */

import {
  smtpClient,
  EmailBuilder,
  TlsMode,
  createAttachment,
  createInlineImage,
} from '@integrations/smtp';
import { readFileSync } from 'fs';
import { join } from 'path';

async function main() {
  // Create SMTP client
  const client = smtpClient()
    .host('smtp.example.com')
    .port(587)
    .credentials('user@example.com', 'your-password')
    .tlsMode(TlsMode.StartTls)
    .build();

  try {
    // Example 1: Create attachment from file content
    // Load a PDF file to attach
    const pdfPath = join(__dirname, 'documents', 'report.pdf');
    const pdfContent = readFileSync(pdfPath);

    const pdfAttachment = createAttachment({
      filename: 'quarterly-report.pdf',
      content: pdfContent,
      contentType: 'application/pdf',
    });

    // Example 2: Create attachment from text content
    const csvContent = Buffer.from(`
Name,Email,Status
John Doe,john@example.com,Active
Jane Smith,jane@example.com,Active
    `.trim());

    const csvAttachment = createAttachment({
      filename: 'users.csv',
      content: csvContent,
      contentType: 'text/csv',
    });

    // Example 3: Create inline images for HTML email
    // Inline images are embedded in the HTML using Content-ID (CID)
    const logoPath = join(__dirname, 'images', 'logo.png');
    const logoContent = readFileSync(logoPath);

    const logoImage = createInlineImage({
      filename: 'logo.png',
      content: logoContent,
      contentType: 'image/png',
      contentId: 'logo123', // This CID is referenced in the HTML
    });

    const chartPath = join(__dirname, 'images', 'chart.png');
    const chartContent = readFileSync(chartPath);

    const chartImage = createInlineImage({
      filename: 'chart.png',
      content: chartContent,
      contentType: 'image/png',
      contentId: 'chart456',
    });

    // HTML content with inline images
    const htmlContent = `
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <style>
    body {
      font-family: Arial, sans-serif;
      max-width: 600px;
      margin: 0 auto;
      padding: 20px;
    }
    .header {
      text-align: center;
      margin-bottom: 30px;
    }
    .logo {
      max-width: 200px;
      height: auto;
    }
    .chart {
      width: 100%;
      max-width: 500px;
      height: auto;
      margin: 20px 0;
    }
  </style>
</head>
<body>
  <div class="header">
    <!-- Reference inline image using cid: scheme -->
    <img src="cid:logo123" alt="Company Logo" class="logo">
    <h1>Quarterly Report</h1>
  </div>

  <p>Dear Team,</p>

  <p>Please find attached the quarterly report for Q4 2025.</p>

  <h2>Performance Overview</h2>

  <!-- Reference another inline image -->
  <img src="cid:chart456" alt="Performance Chart" class="chart">

  <p>Key highlights:</p>
  <ul>
    <li>Revenue increased by 23%</li>
    <li>Customer satisfaction score: 9.2/10</li>
    <li>New customers: 1,247</li>
  </ul>

  <p>The full report is attached as a PDF. You'll also find a CSV file with the user statistics.</p>

  <p>Best regards,<br>The Analytics Team</p>
</body>
</html>
    `.trim();

    const textContent = `
Quarterly Report

Dear Team,

Please find attached the quarterly report for Q4 2025.

Key highlights:
- Revenue increased by 23%
- Customer satisfaction score: 9.2/10
- New customers: 1,247

The full report is attached as a PDF. You'll also find a CSV file with the user statistics.

Best regards,
The Analytics Team
    `.trim();

    // Build the email with attachments and inline images
    const email = new EmailBuilder()
      .from('analytics@example.com')
      .to('team@example.com')
      .subject('Q4 2025 Quarterly Report')
      .text(textContent)
      .html(htmlContent)
      // Add regular attachments
      .attachment(pdfAttachment)
      .attachment(csvAttachment)
      // Add inline images
      .inlineImage(logoImage)
      .inlineImage(chartImage)
      .build();

    console.log('Sending email with attachments...');
    console.log('Attachments:', email.attachments.length);
    console.log('Inline images:', email.inlineImages.length);

    // Calculate approximate email size
    const totalSize = email.attachments.reduce((sum, att) => sum + att.content.length, 0) +
                     email.inlineImages.reduce((sum, img) => sum + img.content.length, 0);
    console.log(`Total attachment size: ${(totalSize / 1024).toFixed(2)} KB`);

    // Send the email
    const result = await client.send(email);

    console.log('Email sent successfully!');
    console.log('Message ID:', result.messageId);

    // Note: Large attachments may take longer to send
    // Consider using connection pooling for batch operations

  } catch (error) {
    console.error('Failed to send email with attachments:', error);
    throw error;
  } finally {
    await client.close();
  }
}

// Alternative approach: Send multiple emails with attachments efficiently
async function batchSendWithAttachments() {
  // Enable connection pooling for better performance when sending multiple emails
  const client = smtpClient()
    .host('smtp.example.com')
    .port(587)
    .credentials('user@example.com', 'your-password')
    .tlsMode(TlsMode.StartTls)
    .poolConfig({
      maxConnections: 5,
      minIdleConnections: 2,
      maxIdleTime: 60000,
      acquireTimeout: 10000,
    })
    .build();

  try {
    const recipients = [
      'user1@example.com',
      'user2@example.com',
      'user3@example.com',
    ];

    // Prepare a common attachment
    const attachment = createAttachment({
      filename: 'welcome.pdf',
      content: Buffer.from('PDF content here...'),
      contentType: 'application/pdf',
    });

    const emails = recipients.map(recipient =>
      new EmailBuilder()
        .from('sender@example.com')
        .to(recipient)
        .subject('Welcome Package')
        .text('Please find the welcome package attached.')
        .attachment(attachment)
        .build()
    );

    console.log(`Sending ${emails.length} emails with attachments...`);

    // Send all emails
    const result = await client.sendBatch(emails);

    console.log('Batch send complete!');
    console.log('Successful:', result.successful.length);
    console.log('Failed:', result.failed.length);

    if (result.failed.length > 0) {
      console.error('Failed emails:');
      for (const failed of result.failed) {
        console.error(`  - ${failed.email.to[0].address}: ${failed.error.message}`);
      }
    }

  } finally {
    await client.close();
  }
}

// Run the main example
if (require.main === module) {
  main().catch((error) => {
    console.error('Fatal error:', error);
    process.exit(1);
  });
}

// Export for potential reuse
export { main, batchSendWithAttachments };

File: smtp/rust/examples/attachments.rs
//! Email with Attachments Example
//!
//! This example demonstrates how to:
//! - Attach files to an email
//! - Include inline images in HTML body
//! - Handle different attachment types (documents, images, etc.)

use smtp_integration::{
    SmtpClient, SmtpConfig, Email, EmailAddress, Attachment, SmtpError,
};
use std::fs;

#[tokio::main]
async fn main() -> Result<(), SmtpError> {
    // Configure the SMTP client
    let config = SmtpConfig::builder()
        .host("smtp.example.com")
        .port(587)
        .username("user@example.com")
        .password("your-password")
        .use_tls(true)
        .use_starttls(true)
        .build()?;

    // Create the SMTP client
    println!("Creating SMTP client...");
    let client = SmtpClient::new(config)?;

    // Plain text version
    let text_body = r#"
Monthly Report - December 2025

Please find attached the monthly report for December 2025.

The report includes:
- Sales summary
- Performance metrics
- Regional breakdowns

If you have any questions about the report, please don't hesitate to reach out.

Best regards,
Analytics Team
"#.trim();

    // HTML version with inline image reference
    let html_body = r#"
<!DOCTYPE html>
<html>
<head>
    <style>
        body {
            font-family: Arial, sans-serif;
            line-height: 1.6;
            color: #333;
            max-width: 600px;
            margin: 0 auto;
        }
        .header {
            background-color: #2196F3;
            color: white;
            padding: 20px;
            text-align: center;
        }
        .content {
            padding: 20px;
        }
        .logo {
            max-width: 200px;
            margin: 20px auto;
            display: block;
        }
        .highlights {
            background-color: #f0f8ff;
            padding: 15px;
            margin: 15px 0;
            border-radius: 5px;
        }
    </style>
</head>
<body>
    <div class="header">
        <h1>Monthly Report - December 2025</h1>
    </div>
    <div class="content">
        <!-- Inline image reference using Content-ID -->
        <img src="cid:company-logo" alt="Company Logo" class="logo">

        <p>Please find attached the monthly report for December 2025.</p>

        <div class="highlights">
            <h3>Report Contents:</h3>
            <ul>
                <li>Sales summary</li>
                <li>Performance metrics</li>
                <li>Regional breakdowns</li>
            </ul>
        </div>

        <p>If you have any questions about the report, please don't hesitate to reach out.</p>
        <p>Best regards,<br><strong>Analytics Team</strong></p>
    </div>
</body>
</html>
"#.trim();

    // Create a sample PDF attachment (in real use, you'd read an actual file)
    // For demonstration, we'll create a simple text file as an attachment
    let report_content = b"Monthly Report - December 2025\n\nSales: $1,234,567\nGrowth: +15%\n";
    let report_attachment = Attachment::builder()
        .filename("monthly_report_dec2025.pdf")
        .content_type("application/pdf")
        .data(report_content.to_vec())
        .build()?;

    // Create a CSV attachment
    let csv_content = b"Date,Sales,Region\n2025-12-01,50000,North\n2025-12-02,45000,South\n";
    let csv_attachment = Attachment::builder()
        .filename("sales_data.csv")
        .content_type("text/csv")
        .data(csv_content.to_vec())
        .build()?;

    // Create an inline image attachment
    // This is a simple 1x1 PNG (in real use, you'd read an actual logo file)
    let logo_data = vec![
        0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A, // PNG signature
        // ... (simplified for example)
    ];
    let logo_attachment = Attachment::builder()
        .filename("logo.png")
        .content_type("image/png")
        .data(logo_data)
        .content_id("company-logo") // Content-ID for inline reference
        .inline(true) // Mark as inline
        .build()?;

    // Alternative: Read attachments from files
    // Uncomment these lines if you have actual files to attach
    /*
    let report_data = fs::read("path/to/monthly_report.pdf")
        .map_err(|e| SmtpError::IoError(e))?;
    let report_attachment = Attachment::builder()
        .filename("monthly_report_dec2025.pdf")
        .content_type("application/pdf")
        .data(report_data)
        .build()?;

    let logo_data = fs::read("path/to/company_logo.png")
        .map_err(|e| SmtpError::IoError(e))?;
    let logo_attachment = Attachment::builder()
        .filename("logo.png")
        .content_type("image/png")
        .data(logo_data)
        .content_id("company-logo")
        .inline(true)
        .build()?;
    */

    // Build the email with attachments
    let email = Email::builder()
        .from(EmailAddress::new("analytics@example.com", Some("Analytics Team")))
        .to(EmailAddress::new("manager@example.com", Some("Department Manager")))
        .cc(EmailAddress::new("director@example.com", Some("Director")))
        .subject("Monthly Report - December 2025")
        .text_body(text_body)
        .html_body(html_body)
        .attachment(report_attachment)
        .attachment(csv_attachment)
        .attachment(logo_attachment)
        .build()?;

    println!("Sending email with attachments...");
    println!("  From: {}", email.from.email);
    println!("  To: {}", email.to.iter().map(|a| a.email.as_str()).collect::<Vec<_>>().join(", "));
    println!("  Subject: {}", email.subject);
    println!("  Attachments: {}", email.attachments.len());

    for attachment in &email.attachments {
        if attachment.inline {
            println!("    - {} (inline, Content-ID: {})",
                attachment.filename,
                attachment.content_id.as_deref().unwrap_or("N/A"));
        } else {
            println!("    - {} ({} bytes)",
                attachment.filename,
                attachment.data.len());
        }
    }

    // Send the email
    match client.send(&email).await {
        Ok(response) => {
            println!("Email sent successfully!");
            println!("  Message ID: {:?}", response.message_id);
            println!("  Status: {}", response.status);
        }
        Err(e) => {
            eprintln!("Failed to send email: {}", e);
            return Err(e);
        }
    }

    Ok(())
}
